zk/stages/stage_batches.go

```diff
@@ -890,3 +906,54 @@ func writeL2Block(eriDb ErigonDb, hermezDb HermezDb, l2Block *types.FullL2Block,
 
 	return nil
 }
+
+// findAncestorBlockNumber searches the latest ancestor block from the database, that matches specified hash using the binary search.
+// Parameters:
+//   - db: An instance of ErigonDb, which provides methods to read data from the blockchain database.
+//   - dsClient: An instance of DatastreamClient, that provides mechanism to read data from the sequencer.
+//   - latestBlockNum: The upper bound of search interval.
+//   - targetHash: The hash of the block that the function is trying to find.
+//
+// Returns:
+//   - uint64: The block number of the latest ancestor block encountered with the specified hash.
+//   - error: An error object that is not nil if the function encounters an issue while reading the
+//     database or if the target hash is not found.
+func findAncestorBlockNumber(db *erigon_db.ErigonDb, dsClient DatastreamClient, latestBlockNum uint64, targetHash common.Hash) (uint64, error) {
+	var (
+		startBlockNum = uint64(0)
+		endBlockNum   = latestBlockNum
+		blockNumber   *uint64
+	)
+
+	if startBlockNum >= endBlockNum {
+		return 0, fmt.Errorf("failed to find ancestor with hash %s in the db."+
+			"The provided block interval is invalid (start block=%d, end block=%d).",
+			targetHash, startBlockNum, latestBlockNum)
+	}
+
+	for startBlockNum < endBlockNum {
+		midBlockNum := (startBlockNum + endBlockNum) / 2
+		midBlockStream, err := dsClient.GetL2BlockByNumber(midBlockNum)
+		if err != nil {
+			return 0, err
+		}
+
+		midBlockDbHash, err := db.ReadCanonicalHash(midBlockNum)
+		if err != nil {
+			return 0, err
+		}
+
+		if midBlockStream.L2Blockhash == midBlockDbHash {
```

Comment:
It occurs to me that in a resequence event, a block's parent hash can stay the same, but the batch number of the block could be different. Could you please add a condition check to verify the batch number of a block also matches the datastream? The batch number stored for local block could be fetched by `hermezDb.GetBatchNoByL2Block(blockNo)`.

Reply:
With the addition of checking batch number, we need to also check if the batch number of an already processed block matches to the data stream [here](https://github.com/0xPolygonHermez/cdk-erigon/blob/92e7d22534bc103203a4a699b8de60a4a5df4c3e/zk/stages/stage_batches.go#L287-L291).

Reply:
> It occurs to me that in a resequence event, a block's parent hash can stay the same, but the batch number of the block could be different. Could you please add a condition check to verify the batch number of a block also matches the datastream? The batch number stored for local block could be fetched by hermezDb.GetBatchNoByL2Block(blockNo).

Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/ea58d697ed124dc75afaaf9659f515582dff7f0f.

Reply:
> With the addition of checking batch number, we need to also check if the batch number of an already processed block matches to the data stream [here](https://github.com/0xPolygonHermez/cdk-erigon/blob/92e7d22534bc103203a4a699b8de60a4a5df4c3e/zk/stages/stage_batches.go#L287-L291).

Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/2a3b9c70438ee4d9d4c3fc963e89e15a1d3bd314